### PR TITLE
Join parallelization Spark changes

### PR DESCRIPTION
**Pull request by ezvz**
## Summary

- Adds spark execution path for just running the `left` side, which includes bootstrap computation.
- Adds spark execution path for just running the `finalJoin`, which assumes all joinPart and left side tables are populated.
- Removes logic to only save joinPart tables if not noAgg. Now we also save a table for noAgg case.

## Why / Goal
Will be used in parallelized join exeuction.

## Test Plan
So far I tested this in docker with:

```
CHRONON_DRIVER_JAR=/srv/chronon_jar/spark/target/scala-2.12/spark_uber-assembly-vz--join_parallelization-0.0.71-SNAPSHOT.jar /srv/chronon_jar/api/py/ai/chronon/repo/run.py --conf production/joins/quickstart/training_set.v1 --mode backfill-left

CHRONON_DRIVER_JAR=/srv/chronon_jar/spark/target/scala-2.12/spark_uber-assembly-vz--join_parallelization-0.0.71-SNAPSHOT.jar /srv/chronon_jar/api/py/ai/chronon/repo/run.py --conf production/joins/quickstart/training_set.v1 --mode backfill --selected-join-parts quickstart_purchases_v1 --use-cached-left

CHRONON_DRIVER_JAR=/srv/chronon_jar/spark/target/scala-2.12/spark_uber-assembly-vz--join_parallelization-0.0.71-SNAPSHOT.jar /srv/chronon_jar/api/py/ai/chronon/repo/run.py --conf production/joins/quickstart/training_set.v1 --mode backfill --selected-join-parts quickstart_returns_v1 --use-cached-left

CHRONON_DRIVER_JAR=/srv/chronon_jar/spark/target/scala-2.12/spark_uber-assembly-vz--join_parallelization-0.0.71-SNAPSHOT.jar /srv/chronon_jar/api/py/ai/chronon/repo/run.py --conf production/joins/quickstart/training_set.v1 --mode backfill --selected-join-parts quickstart_users_v1 --use-cached-left

CHRONON_DRIVER_JAR=/srv/chronon_jar/spark/target/scala-2.12/spark_uber-assembly-vz--join_parallelization-0.0.71-SNAPSHOT.jar /srv/chronon_jar/api/py/ai/chronon/repo/run.py --conf production/joins/quickstart/training_set.v1 --mode backfill-final
```

## Reviewers
@donghanz @hzding621

### CODE REVIEW

File: api/py/ai/chronon/repo/run.py [hunk 1]
```diff
@@ -41,6 +41,8 @@
 ]
 SPARK_MODES = [
     "backfill",
+    "backfill-left",
+    "backfill-final",
```
Comment (better365) [inline api/py/ai/chronon/repo/run.py:45]:
```suggestion
    "backfill-join-left",
    "backfill-join-final",
```

File: api/py/ai/chronon/repo/run.py [hunk 2]
```diff
@@ -61,6 +63,8 @@
 
 MODE_ARGS = {
     "backfill": OFFLINE_ARGS,
+    "backfill-left": OFFLINE_ARGS,
+    "backfill-final": OFFLINE_ARGS,
```
Comment (better365) [inline api/py/ai/chronon/repo/run.py:67]:
```suggestion
    "backfill-join-left": OFFLINE_ARGS,
    "backfill-join-final": OFFLINE_ARGS,
```

File: api/py/ai/chronon/repo/run.py [hunk 3]
```diff
@@ -91,6 +95,8 @@
     },
     "joins": {
         "backfill": "join",
+        "backfill-left": "join-left",
+        "backfill-final": "join-final",
```
Comment (better365) [inline api/py/ai/chronon/repo/run.py:99]:
```suggestion
        "backfill-join-left": "join-left",
        "backfill-join-final": "join-final",
```

File: api/py/ai/chronon/repo/run.py [hunk 4]
```diff
@@ -293,10 +299,10 @@ def set_runtime_env(args):
                         .get(effective_mode, {})
                     )
                     # Load additional args used on backfill.
-                    if custom_json(conf_json) and effective_mode == "backfill":
-                        environment["conf_env"]["CHRONON_CONFIG_ADDITIONAL_ARGS"] = (
-                            " ".join(custom_json(conf_json).get("additional_args", []))
-                        )
+                    if custom_json(conf_json) and effective_mode in ["backfill", "backfill-left", "backfill-final"]:
```
Comment (better365) [inline api/py/ai/chronon/repo/run.py:302]:
```suggestion
                    if custom_json(conf_json) and effective_mode in ["backfill", "backfill-join-left", "backfill-join-final"]:
```

File: spark/src/main/scala/ai/chronon/spark/Driver.scala [hunk 5]
```diff
@@ -275,6 +275,68 @@ object Driver {
...
+        with LocalExportTableAbility
+        with ResultValidationAbility {
+      val runFirstHole: ScallopOption[Boolean] =
+        opt[Boolean](required = false,
+                     default = Some(false),
+                     descr = "Skip the first unfilled partition range if some future partitions have been populated.")
+      val startPartitionOverride: ScallopOption[String] =
+        opt[String](required = false,
+                    descr =
+                      "Start date to compute join backfill, this start date will override start partition in conf.")
```
Comment (better365) [inline spark/src/main/scala/ai/chronon/spark/Driver.scala:292]: these two params have been refactored into the offlinecommand. They are not needed any more here. 
```suggestion
```

File: spark/src/main/scala/ai/chronon/spark/Driver.scala [hunk 6]
```diff
@@ -275,6 +275,68 @@ object Driver {
...
```
Comment (better365) [inline spark/src/main/scala/ai/chronon/spark/Driver.scala:323]: Same as above. 
```suggestion
```

File: spark/src/main/scala/ai/chronon/spark/Join.scala [hunk 7]
```diff
@@ -42,8 +42,12 @@ import scala.collection.parallel.ExecutionContextTaskSupport
 import scala.concurrent.duration.{Duration, DurationInt}
 import scala.concurrent.{Await, ExecutionContext, ExecutionContextExecutorService, Future}
 import scala.collection.compat._
+import scala.jdk.CollectionConverters._
+
+import ai.chronon.api.DataModel.Entities
+//import scala.jdk.CollectionConverters.asScalaBufferConverter
```
Comment (better365) [inline spark/src/main/scala/ai/chronon/spark/Join.scala:48]: remove?

File: spark/src/main/scala/ai/chronon/spark/JoinBase.scala [hunk 8]
```diff
@@ -124,76 +128,73 @@ abstract class JoinBase(joinConf: api.Join,
...
+          // and joinTable to be out of sync. skipping behavior is already handled in the outer loop.
+          skipFirstHole = false
+        )
+        .getOrElse(Seq())
+
+      val unfilledRangeCombined = if (!unfilledRanges.isEmpty && smallMode) {
+        // For small mode we want to "un-chunk" the unfilled ranges, because left side can be sparse
+        // in dates, and it often ends up being less efficient to run more jobs in an effort to
+        // avoid computing unnecessary left range. In the future we can look for more intelligent chunking
+        // as an alternative/better way to handle this.
```
Comment (better365) [inline spark/src/main/scala/ai/chronon/spark/JoinBase.scala:156]: great comments!

**Review by better365 — COMMENTED** [review]
The flow looks good. Left a few comments.

File: spark/src/main/scala/ai/chronon/spark/Driver.scala [hunk 9]
```diff
@@ -240,6 +240,11 @@ object Driver {
         with ResultValidationAbility {
       val selectedJoinParts: ScallopOption[List[String]] =
         opt[List[String]](required = false, descr = "A list of join parts that require backfilling.")
+      val useCachedLeft: ScallopOption[Boolean] =
+        opt[Boolean](
+          required = false,
+          default = Some(false),
```
Comment (hzding621) [inline spark/src/main/scala/ai/chronon/spark/Driver.scala:246]: why would this ever be false?

**Review by donghanz — APPROVED** [review]

**Review by hzding621 — APPROVED** [review]

**Pull request merged**